src/Socket.cpp

```diff
@@ -34,57 +35,59 @@ void Socket::write(const std::string& buff) {
     write(buff.c_str(), buff.size());
 }
 
+/*
+ * Reads buffsize bytes. Returns number of bytes read.
+ */
 void Socket::read(char* buff, size_t buffsize) {
-    while(buffsize > 0) {
-        ssize_t res = ::read(_fd, buff, buffsize);
-        if(res == 0) {
-            throw std::runtime_error("Socket::read() - client closed connection");
-        } else if(res < 0) {
-            if(errno == EINTR)
-                continue;
-            throw std::runtime_error("Socket::read() - read() failed");
-        }
+    if(buffsize > MAX_BUFFER_SIZE)
+        throw new std::invalid_argument("Socket::read(): buffsize too large");
 
-        buff += res;
-        buffsize -= res;
-    }
+    while(_bufferSize < buffsize)
+        _bufferedRead();
+
+    memcpy(buff, _buffer, buffsize);
+    _bufferSize -= buffsize;
+    memmove(_buffer, _buffer + buffsize, _bufferSize);
 }
 
 /*
- * Reads a line terminated by \r\n. Returns length of a received string.
- *
- * Current implementation reads bytes one by one. It's probably not the
- * best approach, but it's simple and doesn't require a temporary buffer
- * which should be somehow shifted all the time. So until this method
- * becomes a bottleneck for someone it's good enough.
+ * Reads a line terminated by \r\n or \n. Returns length of a received string.
  */
 size_t Socket::readLine(char* buff, size_t buffsize) {
-    size_t currsize = 0;
-    // TODO: optimize, figure out how many bytes are available using ioctl
-    while(currsize < buffsize) {
-        ssize_t res = ::read(_fd, buff + currsize, 1);
+    size_t end = 0;
+    do {
+        if(end >= buffsize)
+            throw std::runtime_error("Socket::readLine - line too long");
+        if(end == _bufferSize)
+            _bufferedRead();
+        else
+            end++;
+    } while(_buffer[end] != '\n');
```

Comment:
There is a bug here. It will not work for an empty string terminated with \n since you are checking starting with end = 1.